core/src/main/java/com/sequenceiq/cloudbreak/converter/v4/environment/network/SubnetSelector.java

```diff
@@ -17,21 +22,42 @@
 public class SubnetSelector {
     private static final Logger LOGGER = LoggerFactory.getLogger(SubnetSelector.class);
 
-    public Optional<CloudSubnet> chooseSubnet(String preferredSubnetId, Map<String, CloudSubnet> subnetMetas, String availabilityZone, boolean allowRandom) {
-        Optional<CloudSubnet> cloudSubnet;
+    @Inject
+    private EntitlementService entitlementService;
+
+    public Optional<CloudSubnet> chooseSubnet(String preferredSubnetId, Map<String, CloudSubnet> subnetMetas, String availabilityZone,
+            boolean allowRandom) {
+        return chooseSubnet(preferredSubnetId, subnetMetas, availabilityZone, allowRandom, false);
+    }
+
+    @VisibleForTesting
+    Optional<CloudSubnet> chooseSubnetPreferPublic(String preferredSubnetId, Map<String, CloudSubnet> subnetMetas, String availabilityZone) {
```

Comment:
I thought about this a bit more, and think it would be nice to just drop this method and open up the `chooseSubnet` method as package-private and `@VisibleForTesting`.

I think we can get a similar level of readability by replacing the booleans with enums:
```java
public enum SubnetStrategy {
    PREFER_PUBLIC,
    DEFAULT;
}
```

```java
public Optional<CloudSubnet> chooseSubnet(String preferredSubnetId, Map<String, CloudSubnet> subnetMetas, String availabilityZone,
        boolean allowRandom) {
    return chooseSubnet(preferredSubnetId, subnetMetas, availabilityZone, allowRandom, SubnetStrategy.PREFER_PRIVATE);
}

@VisibleForTesting
Optional<CloudSubnet> chooseSubnetPreferPublic(String preferredSubnetId, Map<String, CloudSubnet> subnetMetas, String availabilityZone) {
    return chooseSubnet(preferredSubnetId, subnetMetas, availabilityZone, false, SubnetStrategy.DEFAULT);
}
```

```java
if (strategy == SubnetStrategy.PREFER_PUBLIC) {
```

And then dropping the additional methods.

Reply:
`preferredSubnetId` is always null when it's called through `chooseSubnetPreferPublic`, but is not always null when  it's called through `chooseSubnet`.

Reply:
I can change it to enum to make it more readable. But it's not clear to me what methods you're suggesting dropping. The method signatures in your example have the same accessibility of the existing ones, so I don't understand what's being changed outside of exchanging the boolean value for an enum.

Reply:
Added `SelectionFallbackStrategy` and `SelectionVisibilityStrategy` enums.